tokio/src/sync/rwlock/read_guard.rs

```diff
@@ -0,0 +1,156 @@
+use crate::sync::batch_semaphore::Semaphore;
+use std::fmt;
+use std::marker;
+use std::mem;
+use std::ops;
+
+/// RAII structure used to release the shared read access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`read`] method on
+/// [`RwLock`].
+///
+/// [`read`]: method@crate::sync::RwLock::read
+/// [`RwLock`]: struct@crate::sync::RwLock
+pub struct RwLockReadGuard<'a, T: ?Sized> {
+    pub(super) s: &'a Semaphore,
+    pub(super) data: *const T,
+    pub(super) marker: marker::PhantomData<&'a T>,
```

Comment:
I don't think this is necessary - the semaphore reference causes the guard to be already covariant over `'a`, since all references are covariant over their lifetime, and the data pointer causes the guard to be already covariant over `T`, since const pointers and shared references are covariant over their type. The same applies to the write guards, where `T` is invariant.

Reply:
Is this `PhantomData` really for the variance? I thought it was for dropck or lifetime. https://github.com/tokio-rs/tokio/pull/2472/files#diff-08219750eaa39a07ef0443816d92fb43090dd12b33009efe93ef2db92a48b9b5R135-R136

Reply:
I agree that it isn't necessary (I just copied them from existing code). Dropcheck is only important for non-reference `PhantomData<T>`, `'a` is correctly covariant and `T` is correctly covariant for the read-lock and invariant for the write-lock.

Reply:
> Dropcheck is only important for non-reference `PhantomData<T>`

Oh, you're right. It is only important if it is owned.